transports/janus_pfunix.c

```diff
@@ -0,0 +1,532 @@
+typedef struct janus_pfunix_client {
+	int fd;					/* Client socket */
+	gboolean admin;			/* Whether this client is for the Admin or Janus API */
+	GAsyncQueue *messages;	/* Queue of outgoing messages to push */
+	gint session_timeout:1;	/* Whether a Janus session timeout occurred in the core */
+} janus_pfunix_client;
+static GHashTable *clients = NULL, *clients_by_fd = NULL;
+static janus_mutex clients_mutex;
+
+
+/* Transport implementation */
+int janus_pfunix_init(janus_transport_callbacks *callback, const char *config_path) {
+	if(g_atomic_int_get(&stopping)) {
+		/* Still stopping from before */
+		return -1;
+	}
+	if(callback == NULL || config_path == NULL) {
+		/* Invalid arguments */
+		return -1;
+	}
+
+	/* This is the callback we'll need to invoke to contact the gateway */
+	gateway = callback;
+
+	/* Read configuration */
+	char filename[255];
+	g_snprintf(filename, 255, "%s/%s.cfg", config_path, JANUS_PFUNIX_PACKAGE);
+	JANUS_LOG(LOG_VERB, "Configuration file: %s\n", filename);
+	janus_config *config = janus_config_parse(filename);
+	if(config != NULL) {
+		/* Handle configuration */
+		janus_config_print(config);
+
+		/* First of all Initialize the socketpair for writeable notifications */
+		if(socketpair(PF_LOCAL, SOCK_STREAM, 0, write_fd) < 0) {
+			JANUS_LOG(LOG_FATAL, "Error creating socket pair for writeable events: %d, %s\n", errno, strerror(errno));
+			return -1;
+		}
+
+		/* Setup the Janus API Unix Sockets server(s) */
+		janus_config_item *item = janus_config_get_item_drilldown(config, "general", "enabled");
+		if(!item || !item->value || !janus_is_true(item->value)) {
+			JANUS_LOG(LOG_WARN, "Unix Sockets server disabled (Janus API)\n");
+		} else {
+			item = janus_config_get_item_drilldown(config, "general", "path");
+			char *pfname = (char *)(item && item->value ? item->value : "/tmp/ux-janusapi");
+			if(strlen(pfname) > 108) {
+				JANUS_LOG(LOG_WARN, "The provided path name (%s) is longer than 108 characters, it will be truncated\n", pfname);
+				pfname[108] = '\0';
+			}
+			/* Create socket */
+			pfd = socket(PF_UNIX, SOCK_SEQPACKET, 0);
+			if(pfd < 0) {
+				JANUS_LOG(LOG_FATAL, "Unix Sockets %s creation for Janus API failed: %d, %s\n", pfname, errno, strerror(errno));
+				pfd = 0;
+			} else {
+				/* Unlink before binding */
+				unlink(pfname);
+				/* Let's bind to the provided path now */
+				struct sockaddr_un address;
+				memset(&address, 0, sizeof(address));
+				address.sun_family = AF_UNIX;
+				g_snprintf(address.sun_path, 108, "%s", pfname);
+				JANUS_LOG(LOG_VERB, "Binding Unix Socket %s... (Janus API)\n", pfname);
+				if(bind(pfd, (struct sockaddr *)&address, sizeof(address)) != 0) {
+					JANUS_LOG(LOG_FATAL, "Bind for Unix Socket %s (Janus API) failed: %d, %s\n", pfname, errno, strerror(errno));
+					close(pfd);
+					pfd = 0;
+				} else {
+					JANUS_LOG(LOG_VERB, "Listening on Unix Socket %s... (Janus API)\n", pfname);
+					if(listen(pfd, 5) != 0) {
```

Comment:
Why 5? Not that it's going to be abused but I'd set it to 128, which is the usual default.